stl/src/multprec.cpp

```diff
@@ -1,190 +1,63 @@
 // Copyright (c) Microsoft Corporation.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-// implements multiprecision math for random number generators
+// Once upon a time, this implemented multiprecision math for linear_congruential_engine.
 
 #include <yvals.h>
 
-_STD_BEGIN
-constexpr int _MP_len = 5;
-using _MP_arr         = unsigned long long[_MP_len];
-
-constexpr int shift                 = 64 / 2;
-constexpr unsigned long long mask   = ~(~0ULL << shift);
-constexpr unsigned long long maxVal = mask + 1;
+#include <__msvc_int128.hpp>
+#include <cstdint>
 
-// TRANSITION, ABI: preserved for binary compatibility
-[[nodiscard]] _CRTIMP2_PURE unsigned long long __CLRCALL_PURE_OR_CDECL _MP_Get(
-    _MP_arr u) noexcept { // convert multi-word value to scalar value
-    return (u[1] << shift) + u[0];
-}
+namespace {
+    using _STD _Unsigned128;
 
-static void add(unsigned long long* u, int ulen, unsigned long long* v,
-    int vlen) noexcept { // add multi-word value to multi-word value
-    int i;
-    unsigned long long k = 0;
-    for (i = 0; i < vlen; ++i) { // add multi-word values
-        u[i] += v[i] + k;
-        k = u[i] >> shift;
-        u[i] &= mask;
-    }
-    for (; k != 0 && i < ulen; ++i) { // propagate carry
-        u[i] += k;
-        k = u[i] >> shift;
-        u[i] &= mask;
-    }
-}
+    using _MP_arr = uint64_t[5]; // Stores a 128-bit value in four 32-bit parts.
+    // Each part was uint64_t for intermediate computations (now unused) and the fifth part was always unused.
 
-// TRANSITION, ABI: preserved for binary compatibility
-_CRTIMP2_PURE void __CLRCALL_PURE_OR_CDECL _MP_Add(
-    _MP_arr u, unsigned long long v0) noexcept { // add scalar value to multi-word value
-    unsigned long long v[2];
-    v[0] = v0 & mask;
-    v[1] = v0 >> shift;
-    add(u, _MP_len, v, 2);
-}
+    [[nodiscard]] _Unsigned128 u128_from_mp(_MP_arr _Wx) noexcept {

```

Comment:
I know you'd like to pass such arrays by reference.
And I know it is squirrelly to change the visible functions signature, even if the ABI would not change.
But here why wouldn't you do that.

Reply:
Aha, because this would not work, that's why!
We have rewritten arrays to pointers in call sties, so adding reference here would not be accepted

Reply:
I would like to, but the problem is that in the ABI functions, the parameter is immediately rewritten to be a pointer (this is different than decay). So the helper functions can't take reference-to-array because there is no array.